### PR TITLE
fix shifting of moving* datapoints by one step into the future

### LINKED ISSUE #2632: [BUG] moving* is shifting data
**Describe the bug**
All moving* functions shift the data by one step into the future.

**To Reproduce**
Query the metric itself and additionally with moving* and a window size of 1.
```
/render?
target=carbon.agents.*.cpuUsage&
target=movingMax(carbon.agents.*.cpuUsage, 1)
```
**Expected behavior**
Both series should be equal.

**Screenshots**
grafik

**Environment (please complete the following information):**
 - OS flavor: Ubuntu
 - Graphite-web version 1.1.4

### PR DESCRIPTION
**Pull request by zivillian**
This is the quick fix from the discussion in #2632 

fixes #2632

### CODE REVIEW

**Comment by deniszh** [comment]
Hi @zivillian ,

I'm afraid that's quick and incorrect fix, though... As you can see many function tests are broken now, because of extra 'None' inserted.

**Comment by zivillian** [comment]
If I'm correct, there is no extra `None` introduced, but it's missing and still present in the expectedData lists.
Is my assumption correct, that the original and moved series schould contain the same number of datapoints? If so, I would update the test data to reflect this.

**Comment by deniszh** [comment]
Ah, mea culpa, you're rightm @zivillian 
E.g. https://github.com/graphite-project/graphite-web/blob/97f4d85750b3bff2c478c2d3b98cf4368ca84756/webapp/tests/test_functions.py#L4687
Then tests need to be correcred, indeed

**Comment by zivillian** [comment]
I'll take a look into this, and also add an assertion that the number of datapoints in both series is equal

**Comment by deniszh** [comment]
I'm still bit hesitating to merge this TBH. Opinions? /cc @ploxiln

**Comment by zivillian** [comment]
@ploxiln @deniszh any thoughts?

File: webapp/graphite/render/functions.py [hunk 1]
```diff
@@ -1312,7 +1312,7 @@ def movingWindow(requestContext, seriesList, windowSize, func='average', xFilesF
 
     effectiveXFF = xFilesFactor if xFilesFactor is not None else series.xFilesFactor
 
-    for i in range(windowPoints, len(series)):
+    for i in range(windowPoints + 1, len(series) + 1):
       nonNull = [v for v in series[i - windowPoints:i] if v is not None]
```
Comment (ploxiln) [inline webapp/graphite/render/functions.py:1316]: This is a tricky bit to wrap one's mind around, but I think this is correct.

The highest value `range(windowPoints + 1, len(series) + 1)` returns is `len(series)`, and then just below that's used as the upper bound in the last slice of `series` which ends up with highest index `len(series) - 1`, which is what we expect. 

A minimal demo:

```
>>> series = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
>>> window = 3
>>> for i in range(window+1, len(series)+1):
...     print(i, series[i-window: i])
... 
4 ['b', 'c', 'd']
5 ['c', 'd', 'e']
6 ['d', 'e', 'f']
7 ['e', 'f', 'g']
8 ['f', 'g', 'h']
```

**Review by ploxiln — APPROVED** [review]

**Review by piotr1212 — APPROVED** [review]

**Comment by deniszh** [comment]
Thanks, @ploxiln for reviewing this! That's indeed quite tricky. Merging.
And thanks to @zivillian for issue and PR too, of course!

**Comment by deniszh** [comment]
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------:|
|❌|1.1.x|Cherrypick failed because the selected commit was a merge commit. Please try again by specifying the parent with the `mainline` argument:

> backport --mainline

or:

> backport --mainline <parent-number>

Or refer to the git documentation for more information: https://git-scm.com/docs/git-cherry-pick#Documentation/git-cherry-pick.txt---mainlineparent-number|

To backport manually run: `node scripts/backport --pr 2682`.
For more info read the [Backport documentation](https://github.com/sqren/backport#backport)

**Comment by deniszh** [comment]
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------:|
|❌|1.1.x|Cherrypick failed because the selected commit was a merge commit. Please try again by specifying the parent with the `mainline` argument:

> backport --mainline

or:

> backport --mainline <parent-number>

Or refer to the git documentation for more information: https://git-scm.com/docs/git-cherry-pick#Documentation/git-cherry-pick.txt---mainlineparent-number|

To backport manually run: `node scripts/backport --pr 2682`.
For more info read the [Backport documentation](https://github.com/sqren/backport#backport)

**Comment by deniszh** [comment]
## 💔 Some backports could not be created

| Status | Branch | Result |
|:------:|:------:|:------:|
|❌|1.1.x|Cherrypick failed because the selected commit was a merge commit. Please try again by specifying the parent with the `mainline` argument:

> backport --mainline

or:

> backport --mainline <parent-number>

Or refer to the git documentation for more information: https://git-scm.com/docs/git-cherry-pick#Documentation/git-cherry-pick.txt---mainlineparent-number|

To backport manually run: `node scripts/backport --pr 2682`.
For more info read the [Backport documentation](https://github.com/sqren/backport#backport)

**Pull request merged**